napari/_vispy/layers/labels.py

```diff
@@ -201,269 +130,58 @@ def __init__(
-    ``2**(index + START_TWO_POWER)`` (where ``START_TWO_POWER=5``), that is,
-    the smallest table is close to ``32 * 32``.
-
-    The function then iterates over all combinations of prime numbers from the
-    lists and checks for a combination that has no collisions for the
-    given keys, returning that combination.
-
-    If no combination can be found, returns None.
-
-    Although keys that collide for all table combinations are rare, they are
-    possible: see ``test_collide_keys`` and ``test_collide_keys2``.
-
-    Parameters
-    ----------
-    keys : np.ndarray
-        array of keys to be inserted into the hashmap,
-        used for collision detection
-    first_dim_index : int
-        index for first dimension of PRIME_NUM_TABLE
-    second_dim_index : int
-        index for second dimension of PRIME_NUM_TABLE
-
-    Returns
-    shp : 2-tuple of int, optional
-        If a table shape can be found that has no collisions for the given
-        keys, return that shape. Otherwise, return None.
-    """
-    for fst_size, snd_size in product(
-        PRIME_NUM_TABLE[first_dim_index],
-        PRIME_NUM_TABLE[second_dim_index],
-    ):
-        fst_crd = (keys // snd_size) % fst_size
-        snd_crd = keys % snd_size
-
-        collision_set = set(zip(fst_crd, snd_crd))
-        if len(collision_set) == len(set(keys)):
-            return fst_size, snd_size
-    return None
-
-
-def _get_shape_from_dict(
-    color_dict: Dict[float, Tuple[float, float, float, float]]
-) -> Tuple[int, int]:
-    """Compute the shape of a 2D hashmap based on the keys in `color_dict`.
-
-    This function finds indices for the first and second dimensions of a
-    table in PRIME_NUM_TABLE based on a target load factor of 0.125-0.25,
-    then calls `_get_shape_from_keys` based on those indices.
-
-    This is quite a low load-factor, but, ultimately, the hash table
-    textures are tiny compared to most datasets, so we choose these
-    factors to minimize the chance of collisions and trade a bit of GPU
-    memory for speed.
-    """
-    keys = np.array([x for x in color_dict if x is not None], dtype=np.int64)
-
-    size = len(keys) / MAX_LOAD_FACTOR
-    size_sqrt = sqrt(size)
-    size_log2 = log2(size_sqrt)
-    max_idx = len(PRIME_NUM_TABLE) - 1
-    max_size = PRIME_NUM_TABLE[max_idx][0] ** 2
-    fst_dim = min(max(int(ceil(size_log2)) - START_TWO_POWER, 0), max_idx)
-    snd_dim = min(max(int(round(size_log2, 0)) - START_TWO_POWER, 0), max_idx)
-
-    if len(keys) > max_size:
-        raise MemoryError(
-            f'Too many labels: napari supports at most {max_size} labels, '
-            f'got {len(keys)}.'
+def build_textures_from_dict(
+    color_dict: Dict[int, ColorTuple], max_size: int
+) -> np.ndarray:
+    if len(color_dict) > 2**23:
+        raise ValueError(  # pragma: no cover
+            f"Cannot build a texture with more than 2**23 colors, got {len(color_dict)}"
```

Comment:
```suggestion
    if len(color_dict) > min(2**23, max_size**2):
        raise ValueError(  # pragma: no cover
            f"Cannot build a texture with more than {min(2**23, max_size**2)} colors, got {len(color_dict)}"
```

Reply:
This does not explain if the limitation is from hardware or just from data

Reply:
ah ok I see, though 2**23 would always evaluate to one number so if it is not that than it must be the other option.

Reply:
but in your error you use `min` without information from where the number comes.